Objects/unicodeobject.c

```diff
@@ -15203,13 +15442,32 @@ _PyUnicode_ClearInterned(PyInterpreterState *interp)
         int shared = 0;
         switch (PyUnicode_CHECK_INTERNED(s)) {
         case SSTATE_INTERNED_IMMORTAL:
+            /* Make immortal interned strings mortal again.
+             *
+             * Currently, the runtime is not able to guarantee that it can exit
+             * without allocations that carry over to a future initialization
+             * of Python within the same process. i.e:
+             *   ./python -X showrefcount -c 'import itertools'
+             *   [237 refs, 237 blocks]
+             *
+             * This should remain disabled (`Py_DEBUG` only) until there is a
+             * strict guarantee that no memory will be left after
+             * `Py_Finalize`.
+             */
+#ifdef Py_DEBUG
             // Skip the Immortal Instance check and restore
             // the two references (key and value) ignored
             // by PyUnicode_InternInPlace().
             _Py_SetMortal(s, 2);
+#ifdef Py_REF_DEBUG
+            /* let's be pedantic with the ref total */
+            _Py_IncRefTotal(_PyThreadState_GET());
+            _Py_IncRefTotal(_PyThreadState_GET());
+#endif
 #ifdef INTERNED_STATS
             total_length += PyUnicode_GET_LENGTH(s);
 #endif
+#endif // Py_DEBUG
```

Comment:
Related: PR gh-113601, issues gh-113055 & gh-113190

Does this change address the situation at all?  Does gh-113601 still apply?

Reply:
This guard at least applies to the 3.12 backport.

As to the legacy `_datetimemodule.c`, the crash comes form two immortalizations:

* `_PyUnicode_InternImmortal()` in `dictobject.c` / `PyDict_SetItemString()`
* `PyUnicode_InternImmortal()` in `marshal.c` / `r_object()` / `case TYPE_SHORT_ASCII:`